src/Altinn.Platform/Altinn.Platform.Authentication/Authentication/Repositories/OrganisationRepository.cs

```diff
@@ -17,48 +17,72 @@ namespace Altinn.Platform.Authentication.Repositories
     /// </summary>
     public class OrganisationRepository : IOrganisationRepository
     {
-        private Dictionary<string, Organisation> orgNumberToOrganisation = new Dictionary<string, Organisation>();
-        private Dictionary<string, Organisation> orgToOrganisation = new Dictionary<string, Organisation>();
-        private DateTime dictionaryLastUpdated = DateTime.MinValue;
+        private readonly IMemoryCache _memoryCache;
+        private readonly MemoryCacheEntryOptions _cacheEntryOptions;
 
-        private readonly HttpClient _httpClient = new HttpClient();
-        private readonly ILogger logger;
+        private readonly HttpClient _httpClient;
+        private readonly ILogger _logger;
 
-        private readonly Uri organisationListLocation;
+        private readonly Uri _organisationListLocation;
+
+        private DateTime _lastHarvest;
 
         /// <summary>
         /// Instantiates the class.
         /// </summary>
+        /// <param name="httpClient">the http client</param>
+        /// <param name="memoryCache">the memory cache</param>
         /// <param name="logger">the logger</param>
         /// <param name="generalSettings">the settings which contains the url to the json organisation file</param>
-        public OrganisationRepository(ILogger<OrganisationRepository> logger, IOptions<GeneralSettings> generalSettings)
+        public OrganisationRepository(HttpClient httpClient, IMemoryCache memoryCache, ILogger<OrganisationRepository> logger, IOptions<GeneralSettings> generalSettings)
         {
-            this.logger = logger;
-            organisationListLocation = new Uri(generalSettings.Value.GetOrganisationRepositoryLocation);
+            _httpClient = httpClient;
+            _logger = logger;
+            _organisationListLocation = new Uri(generalSettings.Value.GetOrganisationRepositoryLocation);
+
+            _memoryCache = memoryCache;
+            _cacheEntryOptions = new MemoryCacheEntryOptions()
+                .SetPriority(CacheItemPriority.Normal)
+                .SetAbsoluteExpiration(new TimeSpan(1, 0, 0));
+
+            _lastHarvest = generalSettings.Value.InitialHarvestDateTime;
```

Comment:
I can't find any logic setting a new value for InitialHarvestDateTime. 

As a more general note I don't like us using a settings object to keep track of something like this. 

My advice now is to refactor more of the logic. Don't isolate this to the OrganisationRepository class. Do we actually need all these different ways of performing lookup? Authentication of application owner is the only user of this and that logic probably need only one lookup. Reconsider caching the original dictionary.

The original implementation of this also had an OrganisationController that exposed more lookup variants, but this service is now only used internally.

Reply:
we cannot cache the original dictionary as the key is the org shortname. To lookup the organisation based on the identity from maskinporten we only have organisation number to go on. We can create a single dictionary, but we will still have the case of having to keep track of the the last harvest took place. 

The property was introduced to be able to test that the cache works. Without being able to override _lastHarvest  from the unit test we will always harvest the organisations in the unit tests.